### PR TITLE
Lights

### PR DESCRIPTION
**Pull request by jytang**
Added a Light class, instead of just having a naked ``light_pos`` in every scene (which assumes it's a directional white light).

Three subclasses of Light: DirectionalLight, PointLight, and SpotLight.

Added 4 lights to the test scene for testing purposes.

### CODE REVIEW

**Review by AznCelebration — APPROVED** [review]
Leave a comment

File: src/scene/point_light.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+               float ambient_coeff = 0.2f) // standard 20% ambience
+            : Light(intensity, ambient_coeff, color),
+              default_position(default_position),
+              quadratic(quadratic) {}
+
+    glm::vec3 get_position() {
+        return glm::vec3(transform.get_world_mat() * glm::vec4(default_position, 1.f));
+    }
+
+    float quadratic; // inverse square falloff factor
```
Comment (metakirby5) [inline src/scene/point_light.h]: More descriptive variable name? I'm not even sure what this does.

File: src/scene/scene.h [hunk 2]
```diff
@@ -3,27 +3,36 @@
 #include <GL/glew.h>
 #include <glm/glm.hpp>
 
-#include "btBulletDynamicsCommon.h"
 #include "game_objects/game_object.h"
 #include "shaders/shader.h"
-#include "boost/signals2.hpp"
+#include "scene_info.h"
+
+// These constants should be equivalent to those in the shaders!
```
Comment (metakirby5) [inline src/scene/scene.h:10]: Is it possible to get the shaders to reference these instead? If not, don't worry about it.

File: src/shaders/basic_shader.cpp [hunk 3]
```diff
@@ -36,23 +39,61 @@ void BasicShader::draw(Mesh *mesh, SceneInfo &scene_info, glm::mat4 to_world) {
...
     glBindTexture(GL_TEXTURE_2D, ShaderManager::shadow.shadow_map_tex);
     set_uni("shadow_map", 0); // ID of this texture=0
 
-    // Send lighting uniforms.
-    set_uni("dir_light.direction", -scene_info.light_pos);
-    set_uni("dir_light.color", Color::WHITE.to_vec()); // white light
-    set_uni("dir_light.ambient_coeff", 0.2f);
+    /* LIGHTS */
+    // Send directional lights.
+    std::string uni_prefix = "dir_lights[";
```
Comment (metakirby5) [inline src/shaders/basic_shader.cpp]: This string stuff is kinda 🤔 
Care to leave a comment explaining what's being done?

**Review by metakirby5 — CHANGES_REQUESTED** [review]
Non-graphics stuff looks okay. 👍

File: src/scene/point_light.h (hunk 1 above)
Comment (jytang) [reply src/scene/point_light.h]: My bad. [Quadratic attenuation factor](https://developer.valvesoftware.com/wiki/Constant-Linear-Quadratic_Falloff).

File: src/shaders/basic_shader.cpp (hunk 3 above)
Comment (jytang) [reply src/shaders/basic_shader.cpp]: Need to fill arrays on the shader. The names of those "uniform locations" need to be given by string; for arrays it's in the format ``array_name[idx].field``. So this is just to construct those strings.

File: src/scene/scene.h (hunk 2 above)
Comment (jytang) [reply src/scene/scene.h:10]: Don't think so :(

**Pull request merged**